[c10d] add an API to get the future result(success or failure) of a collective and customize error handling

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138001
* __->__ #137799

Summary:
This PR is trying to let users to know what exact collective call from the python thread is failing, and 
customize their own error handling function, instead of watchdog thread crashing everything.

This is potentially very useful in fault tolerant training, in which we can have in-process restart.
E.g., when an nccl error is detected, users can potentially abort comms, re-init comms and go back to the previous check pointed step and try again, instead of crashing the whole job.

This is to allow users to check the status of each collective call,
using the ivalue::future libs in PT core. This also allows users to
attach its customized failure handling functions by:
work.get_future_result().then(erro_handling_func)

Note that the above call is also non-blocking for CPU thread
Test Plan:
Added a new test: test_get_future_result to verify the workResutl is
correctly propagated to the users

Tags:

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

shuqiangzhang committed:
[c10d] add an API to get the result(success or failure) of a collective
Summary:
This is to allow users to check the status of each collective call,
using the ivalue::future libs in PT core. This also allows users to
attach its customized failure handling functions by:
work.get_future_result().then(erro_handling_func)

Note that the above call is also non-blocking for CPU thread
Test Plan:
Added a new test: test_get_future_result to verify the workResutl is
correctly propagated to the users

Tags:

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -572,6 +574,7 @@ ProcessGroupNCCL::WorkNCCL::WorkNCCL(const WorkNCCL& w)
       numelIn_(w.numelIn_),
       numelOut_(w.numelOut_),
       store_(w.store_),
+      futureWorkResult_(w.futureWorkResult_),
```
increment the ref of the same future

shuqiangzhang committed:
Update on "[c10d] add an API to get the future result(success or failure) of a collective and customize error handling"


Summary:
This PR is trying to let users to know what exact collective call from the python thread is failing, and 
customize their own error handling function, instead of watchdog thread crashing everything.

This is potentially very useful in fault tolerant training, in which we can have in-process restart.
E.g., when an nccl error is detected, users can potentially abort comms, re-init comms and go back to the previous check pointed step and try again, instead of crashing the whole job.

This is to allow users to check the status of each collective call,
using the ivalue::future libs in PT core. This also allows users to
attach its customized failure handling functions by:
work.get_future_result().then(erro_handling_func)

Note that the above call is also non-blocking for CPU thread
Test Plan:
Added a new test: test_get_future_result to verify the workResutl is
correctly propagated to the users

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/Work.hpp:
```diff
@@ -34,6 +34,13 @@ enum class OpType : std::uint8_t {
+enum class WorkResult : std::uint8_t {
+  SUCCESS = 0,
+  FAILURE = 1,
+  UNKNOWN = 100,
```
nit: why jump from 1 to 100?

shuqiangzhang replied:
no special meaning, just followed the enum tradition the line of 34 (enum class OpType) for unknown result.

wconstab commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2985,6 +2990,27 @@ such as `dist.all_reduce(tensor, async_op=True)`.
+                users can use ``fut.wait()`` to blocking wait for the completion of the work and
+                get the WorkResult by ``fut.value()``.
+                Also, users can use ``fut.then(call_back_func)`` to register a callback function to be called
+                when the work is completed, without blocking the current thread.
```
which thread will the callback run on?

shuqiangzhang replied:
Good question, usually the callback would be run on the thread which marks the future as 'complete', e.g., (https://www.internalfb.com/code/fbsource/[9d20cc497d9438a63edfc443cdb319936f008d66]/fbcode/caffe2/aten/src/ATen/core/ivalue_inl.h?lines=962). But if the future is already completed when users registers the callback, the callback function would be ran on the same thread where usr registration happens

shuqiangzhang commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2648,6 +2648,47 @@ def test_nccl_non_blocking_wait_with_barrier(self):
+        if self.rank == 0:
+            work = process_group.allreduce(torch.rand(10).cuda(self.rank))
+            work.wait()
+            result = work.get_future_result().wait()
```
Yes, blocks until the nccl error happens, the users can decide if they want to wait/block the main CPU thread. This mode is useful when TORCH_NCCL_ASYNC_ERROR_HANDLING is 0 (no handling) currently, such that watchdog only monitors/detects timeout but does not 'kill'/'tear down the process. But even watchdog thread kill the process, the behavior is defined that users thread would unblock first (get failure result) and then watchdog thread throws exceptions to crash the process

shuqiangzhang committed:
Update on "[c10d] add an API to get the future result(success or failure) of a collective and customize error handling"


Summary:
This PR is trying to let users to know what exact collective call from the python thread is failing, and 
customize their own error handling function, instead of watchdog thread crashing everything.

This is potentially very useful in fault tolerant training, in which we can have in-process restart.
E.g., when an nccl error is detected, users can potentially abort comms, re-init comms and go back to the previous check pointed step and try again, instead of crashing the whole job.

This is to allow users to check the status of each collective call,
using the ivalue::future libs in PT core. This also allows users to
attach its customized failure handling functions by:
work.get_future_result().then(erro_handling_func)

Note that the above call is also non-blocking for CPU thread
Test Plan:
Added a new test: test_get_future_result to verify the workResutl is
correctly propagated to the users

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] add an API to get the future result(success or failure) of a collective and customize error handling"


Summary:
This PR is trying to let users to know what exact collective call from the python thread is failing, and 
customize their own error handling function, instead of watchdog thread crashing everything.

This is potentially very useful in fault tolerant training, in which we can have in-process restart.
E.g., when an nccl error is detected, users can potentially abort comms, re-init comms and go back to the previous check pointed step and try again, instead of crashing the whole job.

This is to allow users to check the status of each collective call,
using the ivalue::future libs in PT core. This also allows users to
attach its customized failure handling functions by:
work.get_future_result().then(erro_handling_func)

Note that the above call is also non-blocking for CPU thread
Test Plan:
Added a new test: test_get_future_result to verify the workResutl is
correctly propagated to the users

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2648,6 +2648,47 @@ def test_nccl_non_blocking_wait_with_barrier(self):
+        ar_work = process_group.allreduce(torch.rand(10).cuda(self.rank))
+        ar_work.wait()
+        fut = ar_work.get_future_result()
+        # test adding a callback function
```
nit: is the test still for a callback function?

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2648,6 +2648,47 @@ def test_nccl_non_blocking_wait_with_barrier(self):
+    @requires_nccl_version((2, 4, 0), "Need NCCL 2.4+ for error checking")
+    @skip_if_lt_x_gpu(3)
+    def test_get_future_result(self):
+        def print_fut(fut):
+            print("future is ready:")
+            print(fut.value())
```
ideally we can use a callback function to manipulate some values so that we can verify the time when it is called rather than just do the printing.

shuqiangzhang replied:
@pytorchbot merge

shuqiangzhang replied:
Will update in the stacked PR: https://github.com/pytorch/pytorch/pull/138001

shuqiangzhang commented:
@pytorchbot merge

wconstab reviewed (approved):
I think this is good. But can you find the right place to update docs? I think our main doc page should include a section on error-handling modes and describe which modes we encourage.

And the doc should include the detail you answered to one of my questions- the semantics of enabling watchdog for teardown, but also registering a callback. -- The user can use the callback for printing/logging, but still rely on watchdog to tear down the process.

shuqiangzhang commented:
> I think this is good. But can you find the right place to update docs? I think our main doc page should include a section on error-handling modes and describe which modes we encourage.
> 
> And the doc should include the detail you answered to one of my questions- the semantics of enabling watchdog for teardown, but also registering a callback. -- The user can use the callback for printing/logging, but still rely on watchdog to tear down the process.

Sure, Will update. This new API would still needs watchdog thread to call cudaEventDestroy, which is risky in hanging. I am contemplating some further changes. Will update the doc after that

kwen2501 commented:
How does this compare with the `onCompletionHook` logic? 
Is there a chance of consolidation?

kwen2501 commented:
Shall we use "COMPLETE" instead of "SUCCESS" to align with the `isComplete` API? 
There is another API called `isSuccess` but is being deprecated. (We don't want to attest a collective is successful)
